relayer/relays/beacon/header/header.go

```diff
@@ -0,0 +1,266 @@
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/cache"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/header/syncer"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/header/syncer/scale"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/writer"
+)
+
+var ErrFinalizedHeaderUnchanged = errors.New("finalized header unchanged")
+
+type Header struct {
+	cache  *cache.BeaconCache
+	writer *writer.ParachainWriter
+	syncer *syncer.Syncer
+}
+
+func New(writer *writer.ParachainWriter, beaconEndpoint string, slotsInEpoch uint64, epochsPerSyncCommitteePeriod uint64) Header {
+	return Header{
+		cache:  cache.New(),
+		writer: writer,
+		syncer: syncer.New(beaconEndpoint, slotsInEpoch, epochsPerSyncCommitteePeriod),
+	}
+}
+
+func (h *Header) Sync(ctx context.Context, eg *errgroup.Group) (<-chan uint64, <-chan uint64, error) {
+	latestSyncedPeriod, err := h.writer.GetLastSyncedSyncCommitteePeriod()
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetch last sync commitee: %w", err)
+	}
+
+	h.cache.SetLastSyncedSyncCommitteePeriod(latestSyncedPeriod)
+
+	log.WithField("period", latestSyncedPeriod).Info("set cache: last beacon synced sync committee period")
+
+	periodsToSync, err := h.syncer.GetSyncPeriodsToFetch(latestSyncedPeriod)
+	if err != nil {
+		return nil, nil, fmt.Errorf("check sync committee periods to be fetched: %w", err)
+	}
+
+	log.WithFields(log.Fields{
+		"periods": periodsToSync,
+	}).Info("sync committee periods to be synced")
+
+	for _, period := range periodsToSync {
+		err := h.SyncCommitteePeriodUpdate(ctx, period)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	lastFinalizedHeader, err := h.writer.GetLastStoredFinalizedHeader()
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetch last finalized header: %w", err)
+	}
+
+	lastFinalizedSlot, err := h.writer.GetLastStoredFinalizedHeaderSlot()
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetch last finalized header slot: %w", err)
+	}
+
+	h.cache.FinalizedHeaders = append(h.cache.FinalizedHeaders, lastFinalizedHeader)
+
+	log.WithFields(log.Fields{
+		"hash": lastFinalizedHeader,
+		"slot": lastFinalizedSlot,
+	}).Info("set cache: last finalized header")
+
+	log.Info("starting to sync finalized headers")
+
+	ticker := time.NewTicker(time.Second * 20)
+
+	basicChannel := make(chan uint64)
+	incentivizedChannel := make(chan uint64)
+
+	eg.Go(func() error {
+		defer func() {
+			close(basicChannel)
+			close(incentivizedChannel)
+		}()
+		for {
+			lastBlockNumber, err := h.SyncHeaders(ctx)
+			switch {
+			case errors.Is(err, ErrFinalizedHeaderUnchanged):
+				log.WithField("finalized_header", h.cache.LastFinalizedHeader()).Info("finalized header unchanged")
+			case err != nil:
+				log.WithError(err).Error("error while syncing headers")
+
+				return err
+			default:
+				basicChannel <- lastBlockNumber
+				incentivizedChannel <- lastBlockNumber
+			}
+
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
```

Comment:
Should rather return `nil`. This ensures that upon an error, the subsequent `errgroup.Wait()` call doesn't return this `ctx.Err()` but rather the actual error returned by some other gouroutine that failed. 

```suggestion
				return nil
```

Reply:
Cool! Committed 😄